backend/mora/auth/sso.py

```diff
@@ -5,38 +5,272 @@
-from . import tokens
-from .. import settings
 
 basedir = os.path.dirname(__file__)
 
+blueprint = flask.Blueprint('sso', __name__, static_url_path='',
+                            url_prefix='/saml')
+
+
+class AuthException(Exception):
+    pass
+
+
+def _get_saml_settings(app):
+    """Generate the internal config file for OneLogin"""
+    config = app.config
+
+    insecure = config['SAML_IDP_INSECURE']
+    cert_file = config['SAML_CERT_FILE']
+    key_file = config['SAML_KEY_FILE']
+    requests_signed = config['SAML_REQUESTS_SIGNED']
+    saml_idp_metadata_url = config['SAML_IDP_METADATA_URL']
+    saml_idp_metadata_file = config['SAML_IDP_METADATA_FILE']
+
+    if saml_idp_metadata_file:
+        with open(saml_idp_metadata_file, 'r') as idp:
+            remote = OneLogin_Saml2_IdPMetadataParser.parse(idp.read())
+    else:
+        remote = OneLogin_Saml2_IdPMetadataParser.parse_remote(
+            saml_idp_metadata_url,
+            validate_cert=not insecure
+        )
+
+    s = {
+        "debug": True,
+        "sp": {
+            "entityId": flask.url_for('sso.metadata', _external=True),
+            "assertionConsumerService": {
+                "url": flask.url_for('sso.acs', _external=True),
+                "binding": "urn:oasis:names:tc:SAML:2.0:bindings:HTTP-POST"
+            },
+            "singleLogoutService": {
+                "url": flask.url_for('sso.sls', _external=True),
+                "binding": "urn:oasis:names:tc:SAML:2.0:bindings:HTTP-Redirect"
+            },
+        },
+        "security": {
+            "authnRequestsSigned": requests_signed,
+            "logoutRequestSigned": requests_signed
+        }
+    }
+
+    s.setdefault('sp', {}).update(remote.get('sp'))
+    s.setdefault('idp', {}).update(remote.get('idp'))
+
+    if requests_signed:
+        with open(cert_file, 'r') as cf:
+            cert = cf.read()
+        with open(key_file, 'r') as kf:
+            key = kf.read()
+
+        s['sp'].update({
+            "x509cert": cert,
+            "privateKey": key
+        })
+
+    return s
+
 
-def acs(sender, subject, attributes, auth):
-    """Custom assertion consumer service for use with flask_saml"""
-    flask.session[base.TOKEN_KEY] = tokens.pack(auth.xmlstr)
-    flask.session['username'] = attributes[settings.SSO_SAML_USERNAME_ATTR][0]
+def _prepare_flask_request():
+    """Construct OneLogin-friendly request object from Flask request"""
+    # If server is behind proxys or balancers use the HTTP_X_FORWARDED fields
+    url_data = parse.urlparse(flask.request.url)
+    return {
+        'https': 'on' if flask.request.scheme == 'https' else 'off',
+        'http_host': flask.request.host,
+        'server_port': url_data.port,
+        'script_name': flask.request.path,
+        'get_data': flask.request.args.copy(),
+        'post_data': flask.request.form.copy()
+    }
+
+
+def _prepare_saml_auth(func):
+    """Create and initialize OneLogin SAML2 Auth client"""
```

Comment:
“Decorator to…” — otherwise one might reasonably assume that it was a normal function, or perhaps even a contextmanager. Speaking of which… [`contextlib.contextmanager`](https://docs.python.org/3/library/contextlib.html#contextlib.contextmanager) is pretty neat, and ought to work as a decorator as well 🙂